feat: Validate input fields, UI components, navigation flow, and responsiveness

FlowerWineMonk opened the pull request:
This merge includes test coverage for:
- Validation of input fields and associated error messages
- Navigation between main pages: Pre-login page and Login page
- Verification of UI components such as buttons, links, and other key elements
- Ensured responsive behavior of key pages across screen sizes

FlowerWineMonk committed:
Initial skeleton

FlowerWineMonk committed:
Add Login page

FlowerWineMonk committed:
Change package name

FlowerWineMonk committed:
Add LoginPageCredentials and RegisterPageCredentials

FlowerWineMonk committed:
Add Registration Credential Pages

FlowerWineMonk committed:
refactor: move locator constants directly into @FindBy annotations

FlowerWineMonk committed:
refactor: changed names and added spaces to make code more readable

FlowerWineMonk committed:
Add: DriverFactory and WebDriverManager

FlowerWineMonk committed:
Refactor: Changed name fields and removed unnecessary complexities

FlowerWineMonk committed:
Refactor: Changed AbstractPage to BasePage and added additional info for clearance

FlowerWineMonk committed:
Remove tracked files and add to itinore

FlowerWineMonk committed:
Remove: removed tracked files and add to gitignore

FlowerWineMonk committed:
Add to gitignore PreLoginPage.java

FlowerWineMonk committed:
Refactor: changed names and access modifier also added window fullscreen

FlowerWineMonk committed:
Add: cookie handler

FlowerWineMonk committed:
Re-add PreLoginPage

FlowerWineMonk committed:
Refactor: added PageFactory import to fix an issue and changed the name to BasePage

FlowerWineMonk committed:
Removed unnecessary import and commented out captcha

FlowerWineMonk committed:
Add: Tests and did refactoring

FlowerWineMonk committed:
Removed unnecessary spaces

FlowerWineMonk committed:
Removed captchaField

FlowerWineMonk committed:
Added softAsserts and postcondition to verify the url

FlowerWineMonk committed:
Create README.md

satfaat commented on src/test/java/vention/pagesTest/BaseTest.java:
```diff
@@ -0,0 +1,39 @@
+
+public class BaseTest {
+  WebDriver driver;
+  CookiePopupPage cookiePopupPage;
```
When there will be more pages will you add them all into base test?

FlowerWineMonk replied:
No, I will refactor it. Also, all classes will have their own @BeforeMethod. I am currently including it inside BaseTest because I am only testing one class. However, when I include RegistrationFlowTest, I will separate @BeforeMethod and pages from BaseTest; otherwise, it will be not only unreadable but also error-prone.

FlowerWineMonk committed:
Removed unnecessary verifications

FlowerWineMonk committed:
refactor: clean up test structure and fix syntax issues

- Remove unused WebDriver field from BaseTest and add proper cleanup
- Improve test method naming for better clarity
- Fix missing semicolon and typo in assertion messages
- Enhance overall code quality and maintainability

FlowerWineMonk committed:
Add click by locator

FlowerWineMonk committed:
Re-add register page, create profile page, account selection page

FlowerWineMonk committed:
Delete unused Assert

FlowerWineMonk committed:
Complete registration flow test

FlowerWineMonk committed:
Add HomePage and Book flow test

Anassstasi commented on .gitignore:
```diff
@@ -0,0 +1,4 @@
+/target
+config/
+notes.txt
+src/test/java/vention/pagesTest/HomePageLearnMoreTest.java
```
why this test class is ignored?

FlowerWineMonk replied:
Here, I am ignoring it because at that time, it was yet to be completed. Now, it is done and created a new branch to it with all the changes.

Anassstasi commented on pom.xml:
```diff
@@ -7,12 +7,41 @@
+    <dependency>
+      <groupId>org.seleniumhq.selenium</groupId>
+      <artifactId>selenium-java</artifactId>
+      <version>4.33.0</version>
```
Please move all the versions to properties inside pom file

Anassstasi commented on src/main/java/vention/configLoader/ConfigLoader.java:
```diff
@@ -0,0 +1,23 @@
+  private static Properties properties = new Properties();
+
+  static {
+   try {
```
Please fix file styling. Use built-in formatter

Anassstasi replied:
Please use try with resources to work with files

Anassstasi commented on src/main/java/vention/configLoader/ConfigLoader.java:
```diff
@@ -0,0 +1,23 @@
+
+  static {
+   try {
+      FileInputStream fis = new FileInputStream("src/test/config/credentials.properties");
```
Please move file path to a variable

Anassstasi commented on src/main/java/vention/driver/DriverFactory.java:
```diff
@@ -0,0 +1,34 @@
+  public static WebDriver createLocalDriver(String browser) {
+    WebDriver driver = null;
+    switch (browser.toLowerCase()) {
+      case "chrome":
```
Please add a enum for browser types

Anassstasi commented on src/main/java/vention/driver/DriverManager.java:
```diff
@@ -0,0 +1,22 @@
+
+  public static void quitDriver() {
+    if(DriverManager.getDriver() != null) {
+      DriverManager.getDriver().quit();
```
WHy do u use DriverManager.getDriver() inside DriverManager? you cna access driver by getting it from threadLocalDriver

FlowerWineMonk replied:
I am following encapsulation to access private through public methods.

Anassstasi replied:
encapsulation inside the class? encapsulation relates to accessing fields by external classes not inside itself

Anassstasi replied:
DriverManager.getDriver() get a driver from threadLocal collection . Do u have access to this collection from this method? Is it going to have the data you need?

Anassstasi commented on src/main/java/vention/driver/DriverManager.java:
```diff
@@ -0,0 +1,22 @@
+  }
+
+  public static void quitDriver() {
+    if(DriverManager.getDriver() != null) {
```
Same comment as below

Anassstasi commented on src/main/java/vention/pages/AccountSelectionPage.java:
```diff
@@ -0,0 +1,17 @@
+    super();
+  }
+
+  public void clickContinueWithoutTheBritishAirwaysClubBtn() {
```
Method name and button used inside causes confusion. I'd recommend to name the method 'clickContinueButton'

Anassstasi commented on src/main/java/vention/pages/BasePage.java:
```diff
@@ -0,0 +1,44 @@
+import java.time.Duration;
+
+public abstract class BasePage {
+  WebDriver driver;
```
WHy do u need to init driver? DriverManager allows you to access driver from any point in the framework. It's not necessary to init a variable

Anassstasi replied:
The whole point of DriverManager is to get rid of initialising driver variable

Anassstasi commented on src/main/java/vention/pages/BasePage.java:
```diff
@@ -0,0 +1,44 @@
+
+  public BasePage() {
+    this.driver = DriverManager.getDriver();
+    this.wait = new WebDriverWait(driver, Duration.ofSeconds(10));
```
I'd recommend to add the possibility to create waiter with different waiting time a well (use method overloading). It's not necessary to init waiter insideconstructor. You can do that in a separate method in this class.

Anassstasi replied:
ALways there should be ability to select different waiting time amount

Anassstasi commented on src/main/java/vention/pages/BasePage.java:
```diff
@@ -0,0 +1,44 @@
+  }
+
+  protected void waitForVisibilityAndClickByLocator(By locator) {
+    WebElement element = wait.until(ExpectedConditions.elementToBeClickable(locator));
```
Is it possible to pass webelement instead of locator?

FlowerWineMonk replied:
The issue is that the element isn't initialized with FindBy, ot it apperas dynamically, so a locator click was needed.

I tried to do overloading for waitForVisibilityAndClick but encountered issues. I created byLocator which addresses this issue.

Anassstasi commented on src/main/java/vention/pages/CookiePopupPage.java:
```diff
@@ -0,0 +1,17 @@
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+
+public class CookiePopupPage extends BasePage {
```
If it's a popup it's not a page. Feel free to call it Popup. It's enough

Anassstasi commented on src/main/java/vention/pages/CreateProfilePage.java:
```diff
@@ -0,0 +1,76 @@
+  private WebElement selectTitle;
+
+  @FindBy(id = "firstname--inputtext--input")
+  private WebElement firstNameField;
```
Please rename 'field' to 'input'. That's an input, am i right?

Anassstasi commented on src/main/java/vention/pages/CreateProfilePage.java:
```diff
@@ -0,0 +1,76 @@
+  private WebElement lastNameField;
+
+  @FindBy(id = "countrycode")
+  private WebElement selectPhoneNumber;
```
Please add element type to the variable name

Anassstasi commented on src/main/java/vention/pages/CreateProfilePage.java:
```diff
@@ -0,0 +1,76 @@
+  private WebElement phoneNumberField;
+
+  @FindBy(xpath = "//button[@data-testid='register-submit-button']")
+  private WebElement registerBtn;
```
If all other variable use full version of element type - use that everywhere. Please update btn to button

Anassstasi commented on src/main/java/vention/pages/CreateProfilePage.java:
```diff
@@ -0,0 +1,76 @@
+
+  public void selectTitle(String titleValue) {
+    waitForVisibilityAndClick(selectTitle);
+    waitForVisibilityAndClickByLocator(By.xpath("//option[@value='" + titleValue + "']"));
```
Why are you clicking twice?

FlowerWineMonk replied:
selecting title is a dropdown. First i have to click dropdown and afterwards click the title that I would like to choose.

Anassstasi commented on src/main/java/vention/pages/HomePage.java:
```diff
@@ -0,0 +1,81 @@
+
+public class HomePage extends BasePage {
+  @FindBy(id = "fare")
+  private WebElement selectFare;
```
Please add element type to every element.

Anassstasi commented on src/main/java/vention/pages/HomePage.java:
```diff
@@ -0,0 +1,81 @@
+    waitForVisibilityAndClickByLocator(By.xpath("//option[@value='" + fareWay + "']"));
+  }
+
+  public void fillFromField(String fromCity, String fromCityInfo) {
```
I'd recommend to rename the method to 'enterFromDestination' if i underdood the functionality of this method correct

Anassstasi commented on src/main/java/vention/pages/HomePage.java:
```diff
@@ -0,0 +1,81 @@
+    waitForVisibilityAndClickByLocator(By.id(fromCityInfo));
+  }
+
+  public void fillToField(String toCity, String toCityInfo) {
```
Same as above

Anassstasi commented on src/main/java/vention/pages/HomePage.java:
```diff
@@ -0,0 +1,81 @@
+    waitForVisibilityAndClickByLocator(By.id(toCityInfo));
+  }
+
+  public void fillDepartureDateAndClick(String date) {
```
fill and click what? what are we clicking?

please don't forget about single responsibility principle. 1 method - 1 action. 
In that case please call the method 'selectDepartureDate'

Anassstasi commented on src/test/java/vention/pagesTest/BookFlowTest.java:
```diff
@@ -0,0 +1,50 @@
+    softAssert.assertTrue(homePage.isToFieldDisplayed(), "To field should be visible");
+    softAssert.assertTrue(homePage.isDepartureDateFieldDisplayed(), "Departure date field should be visible");
+    softAssert.assertTrue(homePage.isFindFlightsBtnDisplayed(), "Find flights button should be visible");
+
```
I'd recommend to place assertAll after these assertion. What's the point of proceeding the test if those fields are not present?

Anassstasi commented on src/test/java/vention/pagesTest/BookFlowTest.java:
```diff
@@ -0,0 +1,50 @@
+    softAssert.assertTrue(homePage.isFindFlightsBtnDisplayed(), "Find flights button should be visible");
+
+    homePage.selectFare(ConfigLoader.get("fareWay"));
+    homePage.fillFromField(ConfigLoader.get("fromCity"), ConfigLoader.get("fromCityInfo"));
```
Why do you keep test data and config data inside one file? Please separate that. feel free to keep test data insiode test class

FlowerWineMonk replied:
Fixing it

Anassstasi commented on src/test/java/vention/pagesTest/BookFlowTest.java:
```diff
@@ -0,0 +1,50 @@
+    String beforeUrl = DriverManager.getDriver().getCurrentUrl();
+    homePage.clickFindFlightsBtn();
+
+    String afterUrl = DriverManager.getDriver().getCurrentUrl();
```
Please make it final, it shouldn't be changed

Anassstasi commented on src/test/java/vention/pagesTest/BookFlowTest.java:
```diff
@@ -0,0 +1,50 @@
+    homePage.clickFindFlightsBtn();
+
+    String afterUrl = DriverManager.getDriver().getCurrentUrl();
+    softAssert.assertNotEquals(afterUrl, beforeUrl, "User should be redirected to a new page");
```
Please update assertion message to smth more specific, ex. "Incorrect URL after clicking on ... . User should be redirected to ..."

Anassstasi replied:
Also i think it should be a hard assertion. Feel free to combine soft and hard assertions in one test

Anassstasi commented on src/test/java/vention/pagesTest/LoginFlowTest.java:
```diff
@@ -0,0 +1,49 @@
+
+  @BeforeMethod
+  public void beforeMethod() {
+    String url = ConfigLoader.get("pre-login-url");
```
WHy do you have many urls? How are you going to maintain them later? in case you need to add a new test for a new page you will add a new url?

App has one common URL but URIs might be different and they are related to Pages (Page Objects in terms of automation, you can create a method 'openPage' which is common for all pages and navigate to it via that method)

Anassstasi commented on src/test/java/vention/pagesTest/LoginFlowTest.java:
```diff
@@ -0,0 +1,49 @@
+    cookiePopupPage.clickRejectAllBtn();
+
+    preLoginPage = new PreLoginPage();
+    preLoginPage.clickLoginBtn();
```
Why is this placed in the before mehtod? i think that's part of the test as we are testing login

FlowerWineMonk replied:
this is will redirect us to an actual login page. 
somehow the website is designed in such way.

Anassstasi commented on src/test/java/vention/pagesTest/LoginFlowTest.java:
```diff
@@ -0,0 +1,49 @@
+    softAssert.assertTrue(loginPage.isPasswordFieldDisplayed(), "Password field should be visible");
+    softAssert.assertTrue(loginPage.isNextBtnDisplayed(), "Next button should be visible");
+ 
+    loginPage.fillLoginCredentials(ConfigLoader.get("email"), ConfigLoader.get("password"));
```
Please don't store credentails inside config file. Create UserFactory and store that there

Anassstasi commented on src/test/java/vention/pagesTest/LoginFlowTest.java:
```diff
@@ -0,0 +1,49 @@
+
+    String actualUrl = DriverManager.getDriver().getCurrentUrl();
+    String expectedUrl = ConfigLoader.get("home-url");
+    softAssert.assertEquals(actualUrl, expectedUrl, "User should be redirected to the home page after login");
```
Please reviewe assertion types. Not sure all of them should be soft

Anassstasi commented on src/test/java/vention/pagesTest/RegistrationFlowTest.java:
```diff
@@ -0,0 +1,65 @@
+
+    createProfilePage.selectTitle(ConfigLoader.get("titleValue"));
+    createProfilePage.fillNameCredentials(ConfigLoader.get("registrationFirstName"), ConfigLoader.get("registrationLastName"));
+    createProfilePage.selectCountryCode(ConfigLoader.get("code"));
```
Update the way you store test data

Anassstasi reviewed (changes requested):
Please resolve the comments

FlowerWineMonk committed:
refactor: reorganize test framework structure and simplify driver handling

FlowerWineMonk commented on src/main/java/vention/driver/DriverFactory.java:
```diff
@@ -23,12 +24,14 @@ public static WebDriver createLocalDriver(String browser) {
-    WebDriverManager.chromedriver().setup();
-    return new ChromeDriver();
+    ChromeOptions options = new ChromeOptions();
+    options.addArguments("--start-fullscreen");
```
I added this argument because just writing "fullscreen" won't work with my i3 window manager

FlowerWineMonk replied:
I added an option for chrome since it does not work with just "fullscreen" with my i3 window manager

FlowerWineMonk commented on pom.xml:
```diff
@@ -8,24 +8,45 @@
     </dependency>
 
     <dependency>
-      <groupId>io.github.bonigarcia</groupId>
```
Deleted unnecessary dependency since modern selenium does not require webdrivermanager

Anassstasi commented on src/main/java/vention/configLoader/ConfigLoader.java:
```diff
@@ -0,0 +1,23 @@
+
+public class ConfigLoader {
+  private static final Properties properties = new Properties();
+  private static final String CONFIG_FILE_PATH = "src/test/config/credentials.properties";
```
Why have you hardcoded credentials.properties inside ConfigLoader? I am expecting general config variables inside config loader

Credentials are usually managed inside UserFactory or any other similar solution to manage user data.

FlowerWineMonk replied:
forgot to rename it. Credentials only has one thing, it is base-url, that's it.

Anassstasi commented on src/main/java/vention/entity/User.java:
```diff
@@ -0,0 +1,51 @@
+  private String email;
+  private String phone;
+
+  private User(Builder builder) {
```
Lombok also has builder annotation, feel free to use it

Anassstasi commented on src/main/java/vention/entity/UserFactory.java:
```diff
@@ -0,0 +1,24 @@
+            .password(dotenv.get("EXISTING_USER_PASSWORD")).build();
+  }
+
+  public static User createRandomUser() {
```
Let's raname this method to getRandomUser . It's more common to name methods with 'get' inside factory classes

Anassstasi commented on src/main/java/vention/pages/BasePage.java:
```diff
@@ -0,0 +1,73 @@
+    PageFactory.initElements(DriverManager.getDriver(), this);
+  }
+
+  public String getRelativePath() {
```
Why do you need initialization of this class in the BasePage class as it's specific for each page? 

You can keep this method but make it abstract, override in every page class.

FlowerWineMonk replied:
You can keep this method but make it abstract, override in every page class.

At first, I made it abstract but then realized that I won't be using every paths that is inside the website. After this realization, I decided to actually use overriding to not to make every page have an empty "" or some paths that won't get used in the future.

Anassstasi commented on src/main/java/vention/pages/BasePage.java:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  public String getBaseUrl() {
+    return ConfigLoader.get("base-url");
```
Common codename convention for env variables is 'BASE_URL'

Anassstasi commented on src/main/java/vention/pages/BasePage.java:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  public void openPage() {
+    String fullUrl = getBaseUrl() + getRelativePath();
```
I'd recommend not to create a variable if we use the statement once and it's not too large.

Anassstasi commented on src/main/java/vention/pages/PreLoginPage.java:
```diff
@@ -0,0 +1,29 @@
+
+  @Override
+  public String getRelativePath() {
+    return "/nx/b/account/en/usa/account/pre-login";
```
I'd recommend to move the string to a constant variable in this class. This comment relates to all the similar cases.

We don't usually hardcode values inside the methods, we create constant variables for such cases

Anassstasi commented on src/test/java/vention/pagesTest/BookFlowTest.java:
```diff
@@ -0,0 +1,46 @@
+    softAssert.assertAll();
+
+    homePage.selectFare("oneWay");
+    homePage.enterFromDestination(("New York"), ("new-york,-john-f-kennedy-(ny)-(jfk),-usa"));
```
From the test i don't really understand why do we have these 2 arguemnets and i don't understand what the 2nd one means.

Test should be implemented from the perspective of a user, nothing technical, only business logic.

FlowerWineMonk replied:
The thing is the selection is actually a dropdown and just writing the specific airport will throw an error. A user first has to type a city and then the drop down appears from which a user has to select one of the airports (the second argument is an id of the airport).

Anassstasi commented on src/test/java/vention/pagesTest/BookFlowTest.java:
```diff
@@ -0,0 +1,46 @@
+    homePage.selectFare("oneWay");
+    homePage.enterFromDestination(("New York"), ("new-york,-john-f-kennedy-(ny)-(jfk),-usa"));
+    homePage.enterToDestination(("London"), ("london,-gatwick-(lgw),-united-kingdom"));
+    homePage.selectDepartureDate("calendar-day-30/06/2025");
```
Same comment here. User can only enter the date and you have to process that to the needed format inside the selectDepartureDate  method

FlowerWineMonk replied:
The date selection requires the specific calendar button "ID" because that's how users actually interact with the calendar. They can (of course) type the date but in the selenium's case, it just cannot type it accurately without forgetting the first 2 numbers (I do not know why it happens).

Anassstasi commented on src/test/java/vention/pagesTest/BookFlowTest.java:
```diff
@@ -0,0 +1,46 @@
+    softAssert.assertTrue(homePage.isFindFlightsButtonDisplayed(), "Find flights button should be visible");
+    softAssert.assertAll();
+
+    homePage.selectFare("oneWay");
```
Let's not hardcode test data as strings. Please avoid the cases where a string is hardcoded. 
In such cases it's better to create a constant variable and store the needed info there.

Why? If anything gets changes in the string value, in case of constant variable you will have to update that just once in the variable, in case of hardcoded data - in every place it's used.

Anassstasi commented on src/test/java/vention/pagesTest/LoginFlowTest.java:
```diff
@@ -0,0 +1,50 @@
+    SoftAssert softAssert = new SoftAssert();
+    softAssert.assertTrue(loginPage.isEmailInputDisplayed(), "Email input should be visible");
+    softAssert.assertTrue(loginPage.isPasswordInputDisplayed(), "Password input should be visible");
+    softAssert.assertTrue(loginPage.isNextButtonDisplayed(), "Next button should be visible");
```
In terms of assertion messages we create them regarding what's wrong, ex, "Next button is not visible during login"

Anassstasi commented on src/test/java/vention/pagesTest/LoginFlowTest.java:
```diff
@@ -0,0 +1,50 @@
+    softAssert.assertAll();
+ 
+    User existingUser = UserFactory.getExistingUser();
+    loginPage.enterLoginCredentials(existingUser.getEmail(), existingUser.getPassword());
```
I'd recommend to pass User object to the method. It increases maintainability of the method

If in the future you will need to pass new parameters to the method, you will have to add new arguments to the method (with current approach) but if you pass User object as an argument, in case of any new user arguments you will only need to process them inside the method. Method signature won't change

Anassstasi commented on src/test/java/vention/pagesTest/RegistrationFlowTest.java:
```diff
@@ -0,0 +1,67 @@
+    preLoginPage.clickRegisterButton();
+
+    registerPage = new RegisterPage();
+    registerPage.enterRegisterCredentials(ConfigLoader.get("newEmail"), ConfigLoader.get("newPassword"));
```
These steps are not a precondition, that's a test itself.

The main idea of a test is to perform steps and check the result of the steps. 
Ex. Open page, go to registration page, register, check you are registers - that's test steps

Anassstasi reviewed (changes requested):
Left a few comments

FlowerWineMonk committed:
Refactor: use constants, Lombok builder, clearer messages, and method cleanup

Anassstasi commented on .gitignore:
```diff
@@ -0,0 +1,4 @@
+/target
+config/
+notes.txt
+.env
```
You sure this one should be ignored as well? i think env files are necessary for test runs or?

FlowerWineMonk replied:
They are necessary to me and maven runs, even if they are ignored (locally). They are afterall credentials and it's kind of a good practice to put credentials in some secret file where it won't be sent to the github.

Anassstasi commented on src/main/java/vention/pages/HomePage.java:
```diff
@@ -0,0 +1,88 @@
+import org.openqa.selenium.By;
+
+public class HomePage extends BasePage {
+  private static final String HOME_PATH = "/travel/home/public/en_us/";
```
I'd recommend to rename relative path variable to have the same and generic name in every page class, ex. RELATIVE_PATH

It would help with further maintainability

Anassstasi commented on src/test/java/vention/pagesTest/BookFlowTest.java:
```diff
@@ -0,0 +1,54 @@
+    Assert.assertTrue(actualUrl.contains(HOME_URL), "User should be in home page");
+
+    cookiePopup = new CookiePopup();
+    Assert.assertTrue(cookiePopup.isRejectAllButtonDisplayed(), "Cookie popup should be visible during pre-login page with a reject all button");
```
Is rejecting cookies part of the test? why do we need to fail the test if this button is not displayed? 

Shouldn't we reject them if this button is present, if not - we can proceed anyway

FlowerWineMonk replied:
Because cookies popup to end users regardless if they are using automation or not. That's why, I put them into test to stimulate end user behavior.

FlowerWineMonk committed:
Refactor: rename relative paths, skip cookie asserts, proceed if not shown, add .env for tests

Anassstasi commented on src/test/java/vention/pagesTest/LoginFlowTest.java:
```diff
@@ -0,0 +1,62 @@
+
+
+public class LoginFlowTest extends BaseTest {
+  private static final String PRE_LOGIN_URL = "/pre-login";
```
these constant should be part of page object class as they are related to them.

Anassstasi commented on src/test/java/vention/pagesTest/BookFlowTest.java:
```diff
@@ -0,0 +1,56 @@
+  private static final String FROM_CITY_NAME = "New York";
+  private static final String FROM_CITY_CODE = "new-york,-john-f-kennedy-(ny)-(jfk),-usa";
+  private static final String TO_CITY_NAME = "London";
+  private static final String TO_CITY_CODE = "london,-gatwick-(lgw),-united-kingdom";
```
This code is a technical statement. Such technical moments should be hidden. In the test you should only select 'London' and the rest should be handles inside the method. For this enum would be a great solution

@Getter
@AllArgsConstructor
enum CITIES {
LONDON("London", "london,-gatwick-(lgw),-united-kingdom");

private String displayText;
private String code;

}

Anassstasi replied:
Please remember it's not related to any page object. It's common.

FlowerWineMonk committed:
Refactor: use Cities enum for city selection, move page-specific constants and checks into page classes

Anassstasi commented on .idea/.gitignore:
```diff
@@ -0,0 +1,3 @@
+# Default ignored files
+/shelf/
```
Please remove .idea folder from the PR and add it to gitignore

Anassstasi commented on src/main/java/vention/pages/AccountSelectionPage.java:
```diff
@@ -0,0 +1,29 @@
+import org.openqa.selenium.support.FindBy;
+
+public class AccountSelectionPage extends BasePage {
+  private static final String ACCOUNT_SELECTION_URL = "/account-selection";
```
PLease rename this variable to a generic one like on other classes. Please double check all similar classes

Anassstasi commented on src/main/java/vention/pages/CreateProfilePage.java:
```diff
@@ -0,0 +1,85 @@
+import vention.entity.User;
+
+public class CreateProfilePage extends BasePage {
+  private static final String CREATE_PROFILE_URL = "/register";
```
^^

Anassstasi commented on src/main/java/vention/pages/HomePage.java:
```diff
@@ -0,0 +1,95 @@
+
+public class HomePage extends BasePage {
+  private static final String RELATIVE_PATH = "/travel/home/public/en_us/";
+  private static final String HOME_URL = "/home";
```
Can't u use RELATIVE_PATH for the validation in the test?

Anassstasi replied:
Are you sure both this URIs are necessary?

FlowerWineMonk replied:
You are right, I will refactor it

Anassstasi commented on src/main/java/vention/pages/HomePage.java:
```diff
@@ -0,0 +1,95 @@
+    return RELATIVE_PATH;
+  }
+
+  public boolean isOnHomePage() {
```
This method can be called 'isPageOpened()' and moved to BasePage and overriden in every page object. What do u think?

FlowerWineMonk replied:
It is good and makes code more cleaner

Anassstasi commented on src/main/java/vention/pages/LoginPage.java:
```diff
@@ -0,0 +1,49 @@
+import vention.entity.User;
+
+public class LoginPage extends BasePage {
+  private static final String LOGIN_URL = "/login";
```
^^

Anassstasi commented on src/main/java/vention/pages/PreLoginPage.java:
```diff
@@ -0,0 +1,46 @@
+
+public class PreLoginPage extends BasePage {
+  private static final String RELATIVE_PATH = "/nx/b/account/en/usa/account/pre-login";
+  private static final String PRE_LOGIN_URL = "/pre-login";
```
Same comment as above

Anassstasi commented on src/main/java/vention/pages/RegisterPage.java:
```diff
@@ -0,0 +1,49 @@
+import vention.entity.User;
+
+public class RegisterPage extends BasePage {
+  private static final String REGISTER_URL = "/email-validation";
```
^^

Anassstasi commented on src/test/java/vention/pagesTest/RegistrationFlowTest.java:
```diff
@@ -0,0 +1,84 @@
+
+public class RegistrationFlowTest extends BaseTest {
+  private static final String SELECT_TITLE = "Mr";
+  private static final String SELECT_COUNTRY_CODE = "+1";
```
Variable names should not start with a verb.

Anassstasi reviewed (changes requested):
Please resolve the comments

FlowerWineMonk committed:
refactor: clean up code and improve naming conventions

- Remove redundant URI definitions
- Rename variables to follow proper naming conventions
- Replace redundant isOnLoginPage/isOnHomePage/... with generic
isPageOpened in BasePage
- Add .idea/ to .gitignore

Anassstasi reviewed (approved):
LGTM!

zapletin-dk commented on .env:
```diff
@@ -0,0 +1,2 @@
+EXISTING_USER_EMAIL=abc123ABC@gmail.com
+EXISTING_USER_PASSWORD=abc123ABC
```
I would suggest to use User class and User factory rather than hardcode data from the batch file. Something like getExistingUser()

FlowerWineMonk replied:
If I got it right your are saying that I need to put those values inside .email and .password instead of calling them from .env ".email(dotenv.get("EXISTING_USER_EMAIL"))".

However, I am following security feature or at least I was following it. I want to leave it as it is, what do you think?

FlowerWineMonk replied:
deleted .env + dependency

zapletin-dk commented on src/main/java/vention/pages/CookiePopup.java:
```diff
@@ -0,0 +1,21 @@
+import org.openqa.selenium.support.FindBy;
+
+public class CookiePopup extends BasePage {
+  @FindBy(xpath = "//button[@id = 'ensCloseBanner']")
```
Use @ FindBy(id = "ensCloseBanner") it's more useful and understandable

zapletin-dk commented on src/main/java/vention/pages/HomePage.java:
```diff
@@ -0,0 +1,89 @@
+
+  public void enterFromDestination(String fromCity, String fromCityInfo) {
+    waitForVisibilityAndSendKeys(fromInput, fromCity);
+    waitForVisibilityAndClickByLocator(By.id(fromCityInfo), 30);
```
I suggest to make fromCityInfo as WebElement variable via @ FindBy, as in the future it may be reused many times

FlowerWineMonk replied:
Please, look at the Cities enum. I have it there, it is getting used through that enum.

zapletin-dk replied:
Agree

Anassstasi replied:
Keep as it is for now.

zapletin-dk commented on src/main/java/vention/pages/HomePage.java:
```diff
@@ -0,0 +1,89 @@
+
+  public void enterToDestination(String toCity, String toCityInfo) {
+    waitForVisibilityAndSendKeys(toInput, toCity);
+    waitForVisibilityAndClickByLocator(By.id(toCityInfo), 30);
```
The same as for the fromCity

FlowerWineMonk replied:
I am using it in cities enum

zapletin-dk replied:
Yes, got it

zapletin-dk commented on src/main/java/vention/pages/RegisterPage.java:
```diff
@@ -0,0 +1,49 @@
+    super();
+  }
+
+  public void enterRegisterCredentials(User user) {
```
All the staff you are doing with User such as enter any credentials you can move to the base page, as it can be reused in the other pages as well

Anassstasi replied:
This method is part of RegisterPage. It enters registration data and it should be part of related page. BasePage should contain general methods for all pages which are not page-specific. If you need to reuse such methods in other pages - that's breakage of PO logic. In such cases(if such situations happen often) it makes sense to create an additional layer of steps, their goal is to help with such situations

zapletin-dk commented on src/test/java/vention/pagesTest/BookFlowTest.java:
```diff
@@ -0,0 +1,51 @@
+    homePage = new HomePage();
+    homePage.openPage();
+
+    Assert.assertTrue(homePage.isPageOpened(), "User should be in home page");
```
It's not necessary here as you are testing the option to book flight. 
This assertion should be in openHomePageTest

Anassstasi replied:
This assertion makes sense here. What's the point to proceed with the test if home page is not opened?

zapletin-dk commented on src/test/java/vention/pagesTest/BookFlowTest.java:
```diff
@@ -0,0 +1,51 @@
+    loginSoftAssert.assertTrue(homePage.isToInputDisplayed(), "To input should be visible during booking in home page");
+    loginSoftAssert.assertTrue(homePage.isDepartureDateButtonDisplayed(), "Departure date button should be visible during booking in home page");
+    loginSoftAssert.assertTrue(homePage.isFindFlightsButtonDisplayed(), "Find flights button should be visible during booking in home page");
+    loginSoftAssert.assertAll();
```
As for homePage assertion. Is it necessary in this test to verify every component of the user flow? This test is more as complete flight booking for user than verifying steps.

FlowerWineMonk replied:
I will remove them.

zapletin-dk commented on src/test/java/vention/pagesTest/BookFlowTest.java:
```diff
@@ -0,0 +1,51 @@
+    final String beforeUrl = DriverManager.getDriver().getCurrentUrl();
+    homePage.clickFindFlightsButton();
+    final String afterUrl = DriverManager.getDriver().getCurrentUrl();
+    Assert.assertNotEquals(afterUrl, beforeUrl, "User should be redirected to outbound page after booking a flight");
```
I would suggest to hardcode expected result not computing it

Anassstasi replied:
Both approaches are fine. It's not necessary to update

zapletin-dk commented on src/test/java/vention/pagesTest/LoginFlowTest.java:
```diff
@@ -0,0 +1,56 @@
+public class LoginFlowTest extends BaseTest {
+  private CookiePopup cookiePopup;
+  private PreLoginPage preLoginPage;
+  private LoginPage loginPage;
```
You can add all the variables in the BaseTest

Anassstasi replied:
Both approaches are fine.

BUT i'd place PO initialisation inside BeforeClass method. WHat's the point initialising them inside every test?

zapletin-dk commented on src/test/java/vention/pagesTest/RegistrationFlowTest.java:
```diff
@@ -0,0 +1,84 @@
+    homePage = new HomePage();
+    Assert.assertTrue(homePage.isPageOpened(),  "User should be redirected to the home page after registration");
+  }
+}
```
It would be lovely if you can decompose the test into smaller parts and reuse them. It's really hard to read the test and it would be more harder to handle the problems in the future with this test. May be by creating steps or smth else.

Anassstasi replied:
In another PR, not here.

FlowerWineMonk committed:
refactor: improve test maintainability and code organization

- Fix CookiePopup selector: update xpath from 'ensCloseBanner' to
'endCloseBanner'
- Extract common user credential methods to BasePage for reusability
across LoginPage and RegistrationPage
- Move home page assertion to appropriate test method (openHomePageTest
in HomePageTest)
- Remove unnecessary soft assertions in BookFlowTest
- Replace URL comparison with assertTrue contains OUTBOUND_URL for
better readability
- Consolidate test variables in BaseTest for centralized management

FlowerWineMonk committed:
refactor: reorganize test structure and clean up dependencies

- Remove unnecessary .env file and unused library
- Move test methods to their respective page classes (prelogin, login,
register, account selection, create profile)
- Update RegisterFlowTest and LoginFlowTest to call methods from their
corresponding test classes
- Improve test organization and reduce code duplication

zapletin-dk commented on src/test/java/vention/pagesTest/LoginFlowTest.java:
```diff
@@ -1,53 +1,24 @@
-import org.testng.asserts.SoftAssert;
 import org.testng.Assert;
 
 
```
empty line

FlowerWineMonk replied:
Removed it

zapletin-dk commented on src/test/java/vention/pagesTest/LoginPageTest.java:
```diff
@@ -0,0 +1,28 @@
+    loginSoftAssert.assertTrue(loginPage.isNextButtonDisplayed(), "Next button should be visible during login");
+    loginSoftAssert.assertAll();
+ 
+    User existingUser = UserFactory.getExistingUser();
```
You can make getExistingUser as static method

FlowerWineMonk replied:
It is already static. 
"public static User getExistingUser() {}"

zapletin-dk reviewed (approved):
Minor fixes needed

Aziza03 commented:
Approved
